crates/turbo-tasks-memory/src/memory_backend.rs

```diff
@@ -118,54 +121,58 @@ impl MemoryBackend {
         self.memory_tasks.get(*id).unwrap()
     }
 
-    pub fn on_task_might_become_inactive(&self, task: TaskId) {
-        if let Some(gc_queue) = &self.gc_queue {
-            gc_queue.task_might_become_inactive(task);
-        }
-    }
-
-    pub fn on_task_flagged_inactive(&self, task: TaskId, compute_duration: Duration) {
+    pub fn run_gc(
+        &self,
+        idle: bool,
+        _turbo_tasks: &dyn TurboTasksBackendApi<MemoryBackend>,
+    ) -> bool {
         if let Some(gc_queue) = &self.gc_queue {
-            gc_queue.task_flagged_inactive(task, compute_duration);
-        }
-    }
+            let mut did_something = false;
+            loop {
+                let mem_limit = self.memory_limit;
 
-    pub fn run_gc(&self, idle: bool, turbo_tasks: &dyn TurboTasksBackendApi<MemoryBackend>) {
-        if let Some(gc_queue) = &self.gc_queue {
-            const MAX_COLLECT_FACTOR: u8 = u8::MAX / 8;
-
-            let mem_limit = self.memory_limit;
-
-            let usage = turbo_tasks_malloc::TurboMalloc::memory_usage();
-            let target = if idle {
-                mem_limit * 3 / 4
-            } else {
-                mem_limit * 7 / 8
-            };
-            if usage < target {
-                if idle {
-                    // Always run propagation when idle
-                    gc_queue.run_gc(0, self, turbo_tasks);
+                let usage = turbo_tasks_malloc::TurboMalloc::memory_usage();
+                let target = if idle {
+                    mem_limit * 3 / 4
+                } else {
+                    mem_limit * 7 / 8
+                };
```

Comment:
It might be nice to move these potentially tunable values into constants declared next to the other ones, like `PERCENTAGE_TO_COLLECT`.